File: android/app/src/main/java/com/hoi/gesture/EyeTrackingDetector.kt
package com.hoi.gesture

import androidx.camera.core.ImageProxy
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.google.mlkit.vision.face.FaceLandmark

/**
 * Detects eye/head gaze direction using ML Kit Face Detection.
 * Used in the pointing phase of Hoi! to determine player look direction.
 * Requires ML Kit Face Detection dependency:
 *   implementation 'com.google.mlkit:face-detection:16.1.5'
 */

enum class GazeDirection {
    UP, DOWN, LEFT, RIGHT, CENTER, UNKNOWN;

    val displayName: String get() = when (this) {
        UP      -> "⬆️ Up"
        DOWN    -> "⬇️ Down"
        LEFT    -> "⬅️ Left"
        RIGHT   -> "➡️ Right"
        CENTER  -> "🎯 Center"
        UNKNOWN -> "❓ Unknown"
    }
}

interface EyeTrackingListener {
    fun onDirectionDetected(direction: GazeDirection)
    fun onCalibrationComplete()
}

class EyeTrackingDetector {

    var listener: EyeTrackingListener? = null
    var detectionThreshold: Float = 20f   // pixels offset from baseline

    // Calibration state
    private var isCalibrating = false
    private var calibrationSamples = mutableListOf<Pair<Float, Float>>()
    private var baselineX: Float = 0f
    private var baselineY: Float = 0f
    private val requiredCalibrationSamples = 30

    // Consecutive frame confirmation
    private var consecutiveFrames = 0
    private val requiredFrames = 3
    private var lastDirection = GazeDirection.CENTER

    private val faceDetector = FaceDetection.getClient(
        FaceDetectorOptions.Builder()
            .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
            .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_ALL)
            .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_NONE)
            .build()
    )

    // MARK: - Calibration

    fun beginCalibration() {
        isCalibrating = true
        calibrationSamples.clear()
        println("[EyeTrackingDetector] Calibration started — ask user to look straight ahead")
    }

    private fun finishCalibration() {
        if (calibrationSamples.isEmpty()) return
        baselineX = calibrationSamples.map { it.first }.average().toFloat()
        baselineY = calibrationSamples.map { it.second }.average().toFloat()
        isCalibrating = false
        calibrationSamples.clear()
        println("[EyeTrackingDetector] Calibration complete. Baseline: ($baselineX, $baselineY)")
        listener?.onCalibrationComplete()
    }

    // MARK: - Detection

    @androidx.camera.core.ExperimentalGetImage
    fun detect(imageProxy: ImageProxy) {
        val mediaImage = imageProxy.image ?: run {
            imageProxy.close()
            return
        }

        val inputImage = InputImage.fromMediaImage(mediaImage, imageProxy.imageInfo.rotationDegrees)

        faceDetector.process(inputImage)
            .addOnSuccessListener { faces ->
                if (faces.isNotEmpty()) {
                    processFace(faces[0])
                }
                imageProxy.close()
            }
            .addOnFailureListener {
                println("[EyeTrackingDetector] ML Kit error: ${it.message}")
                imageProxy.close()
            }
    }

    private fun processFace(face: Face) {
        val leftEye  = face.getLandmark(FaceLandmark.LEFT_EYE)
        val rightEye = face.getLandmark(FaceLandmark.RIGHT_EYE)

        val eyeX: Float
        val eyeY: Float

        when {
            leftEye != null && rightEye != null -> {
                eyeX = (leftEye.position.x + rightEye.position.x) / 2f
                eyeY = (leftEye.position.y + rightEye.position.y) / 2f
            }
            leftEye != null -> {
                eyeX = leftEye.position.x
                eyeY = leftEye.position.y
            }
            rightEye != null -> {
                eyeX = rightEye.position.x
                eyeY = rightEye.position.y
            }
            else -> return
        }

        if (isCalibrating) {
            calibrationSamples.add(Pair(eyeX, eyeY))
            if (calibrationSamples.size >= requiredCalibrationSamples) {
                finishCalibration()
            }
            return
        }

        // Live direction detection
        val deltaX = eyeX - baselineX
        val deltaY = eyeY - baselineY
        val absX = Math.abs(deltaX)
        val absY = Math.abs(deltaY)

        val direction = when {
            absX < detectionThreshold && absY < detectionThreshold -> GazeDirection.CENTER
            absY > absX -> if (deltaY < 0) GazeDirection.UP else GazeDirection.DOWN
            else        -> if (deltaX > 0) GazeDirection.RIGHT else GazeDirection.LEFT
        }

        if (direction == lastDirection) {
            consecutiveFrames++
        } else {
            consecutiveFrames = 1
            lastDirection = direction
        }

        if (consecutiveFrames == requiredFrames &&
            direction != GazeDirection.CENTER &&
            direction != GazeDirection.UNKNOWN) {
            listener?.onDirectionDetected(direction)
            consecutiveFrames = 0
        }
    }
}

File: ios/HoiApp/Sources/GestureRecognition/EyeTrackingDetector.swift
import ARKit
import SceneKit

/// Detects eye/head direction using ARKit Face Tracking.
/// Requires TrueDepth front camera (iPhone X and later).
/// Used in the pointing phase of Hoi! to determine which direction the player looks.

enum GazeDirection {
    case up, down, left, right, center, unknown
}

protocol EyeTrackingDelegate: AnyObject {
    func eyeTrackingDidDetectDirection(_ direction: GazeDirection)
    func eyeTrackingCalibrationComplete(baseline: simd_float3)
}

class EyeTrackingDetector: NSObject {

    // MARK: - Properties
    weak var delegate: EyeTrackingDelegate?

    private var arSession: ARSession?
    private var calibrationBaseline: simd_float3?
    private var isCalibrating = false
    private var calibrationSamples: [simd_float3] = []

    // Threshold for direction detection (tuned post-calibration)
    // Adjustable per user during calibration
    var detectionThreshold: Float = 0.12

    // Consecutive frame confirmation (avoids false positives)
    private var consecutiveFrames = 0
    private let requiredFrames = 3
    private var lastDirection: GazeDirection = .center

    // MARK: - ARKit Session

    func start() {
        guard ARFaceTrackingConfiguration.isSupported else {
            print("[EyeTrackingDetector] ARKit face tracking not supported on this device")
            return
        }

        let config = ARFaceTrackingConfiguration()
        config.isLightEstimationEnabled = false

        arSession = ARSession()
        arSession?.delegate = self
        arSession?.run(config, options: [.resetTracking, .removeExistingAnchors])
        print("[EyeTrackingDetector] ARKit session started")
    }

    func stop() {
        arSession?.pause()
        arSession = nil
    }

    // MARK: - Calibration

    /// Call this to begin calibration. Ask user to look straight ahead.
    /// Collects 30 frames (~0.5s at 60fps) to build a baseline.
    func beginCalibration() {
        isCalibrating = true
        calibrationSamples = []
        print("[EyeTrackingDetector] Calibration started — ask user to look straight ahead")
    }

    private func finishCalibration() {
        guard !calibrationSamples.isEmpty else { return }
        // Average all samples to get baseline center gaze vector
        let sum = calibrationSamples.reduce(simd_float3(0, 0, 0), +)
        calibrationBaseline = sum / Float(calibrationSamples.count)
        isCalibrating = false
        calibrationSamples = []
        print("[EyeTrackingDetector] Calibration complete. Baseline: \(calibrationBaseline!)")
        delegate?.eyeTrackingCalibrationComplete(baseline: calibrationBaseline!)
    }

    // MARK: - Direction Classification

    private func classifyGaze(_ lookAtPoint: simd_float3) -> GazeDirection {
        guard let baseline = calibrationBaseline else { return .unknown }

        let delta = lookAtPoint - baseline

        let absX = abs(delta.x)
        let absY = abs(delta.y)

        // Must exceed threshold to register as a direction
        guard absX > detectionThreshold || absY > detectionThreshold else {
            return .center
        }

        // Dominant axis wins
        if absY > absX {
            return delta.y > 0 ? .up : .down
        } else {
            return delta.x > 0 ? .right : .left
        }
    }
}

// MARK: - ARSessionDelegate
extension EyeTrackingDetector: ARSessionDelegate {

    func session(_ session: ARSession, didUpdate anchors: [ARAnchor]) {
        guard let faceAnchor = anchors.first(where: { $0 is ARFaceAnchor }) as? ARFaceAnchor else {
            return
        }

        let lookAt = faceAnchor.lookAtPoint

        if isCalibrating {
            calibrationSamples.append(lookAt)
            if calibrationSamples.count >= 30 {
                finishCalibration()
            }
            return
        }

        // Live detection
        let direction = classifyGaze(lookAt)

        if direction == lastDirection {
            consecutiveFrames += 1
        } else {
            consecutiveFrames = 1
            lastDirection = direction
        }

        // Only fire event after N consecutive matching frames
        if consecutiveFrames == requiredFrames, direction != .center, direction != .unknown {
            DispatchQueue.main.async {
                self.delegate?.eyeTrackingDidDetectDirection(direction)
            }
            consecutiveFrames = 0
        }
    }

    func session(_ session: ARSession, didFailWithError error: Error) {
        print("[EyeTrackingDetector] ARKit session error: \(error)")
    }
}

// MARK: - GazeDirection Helper
extension GazeDirection {
    var displayName: String {
        switch self {
        case .up:      return "⬆️ Up"
        case .down:    return "⬇️ Down"
        case .left:    return "⬅️ Left"
        case .right:   return "➡️ Right"
        case .center:  return "🎯 Center"
        case .unknown: return "❓ Unknown"
        }
    }
}
